### ISSUE TITLE
Spike on introducing react-testing-library as a recommending testing solution

### ISSUE THREAD

**Issue by bjankord** [cerner/terra-toolkit#340]
# Description
Spike on introducing react-testing-library as a recommending testing solution

Things we should explore in this spike:
* How to test react-intl based stuff (translations)
* How to test react-hooks
* Examples available for testing common components, use cases, and workflows
* Migration guide & guidance is available for app teams to refer to

**Comment by cody-dot-js** [comment]
The whole purpose of react testing library is:

> [The more your tests resemble the way your software is used, the more confidence they can give you.](https://github.com/testing-library/react-testing-library#this-solution) 

That being said, [terra-enzyme-intl](https://github.com/cerner/terra-enzyme-intl) is specific for testing code that wrapped in `react-intl` using `enzyme`, but we can use the underlying [logic for making tests independent of actual translations](https://github.com/cerner/terra-enzyme-intl/blob/master/src/index.js#L23-L26) to build something that works well with [react-testing-library](https://github.com/testing-library/react-testing-library) & react-intl. We can also use their guidance on [custom test renderers](https://testing-library.com/docs/react-testing-library/setup#custom-render)

## TLDR:
I made a repo [here](https://github.com/dev-cprice/terra-react-testing-library) demonstrating the above discussion:
* using react-intl (v2, since that's what terra supports)
* using react-testing-library
* using terra components
* using hooks
* using create-react-app, which provides a jest test runner

Testing utils: [here](https://github.com/dev-cprice/terra-react-testing-library/blob/master/src/test-utils.js)
Example component: [here](https://github.com/dev-cprice/terra-react-testing-library/blob/master/src/ExampleComponent.jsx)
Example component test: [here](https://github.com/dev-cprice/terra-react-testing-library/blob/master/src/ExampleComponent.test.js)

Edit:
The repo linked above was a quick attempt at answering the following questions:
* can we use this with terra?
* can we use this with hooks?
* can we use this with react-intl?

Further examples/recipes/investigation to follow

**Comment by cody-dot-js** [comment]
React testing library provides a list of example recipes as well:

* [input events](https://testing-library.com/docs/example-input-event)
* [update props](https://testing-library.com/docs/example-update-props)
* [react redux](https://testing-library.com/docs/example-react-redux)
* [react router](https://testing-library.com/docs/example-react-router)
* [external examples](https://react-testing-examples.com/jest-rtl/)

Other links:
React testing library [docs](https://testing-library.com/docs/react-testing-library/intro)
Example used as source material in the repo linked: [here](https://testing-library.com/docs/react-testing-library/example-intro)
Cheatsheet [here](https://testing-library.com/docs/react-testing-library/cheatsheet), has:
* [queries](https://testing-library.com/docs/react-testing-library/cheatsheet#queries)
* [async](https://testing-library.com/docs/react-testing-library/cheatsheet#async) (like `waitForElement`)
* [events](https://testing-library.com/docs/react-testing-library/cheatsheet#events)

**Comment by dv297** [comment]
Thanks for putting this together. I'm really interested in using React testing library, so I think this spike is a step in the right direction. I've been using it a little bit in some of our projects but I've hit this weird case when consuming terra-alert. I've created a fork of your repository to demonstrate a minimum reproducable case.

https://github.com/dv297/terra-react-testing-library

I've created a "SampleComponent", which just renders a terra-alert along with some text. The text is just helpful for demonstrating some of the weird behavior.

![ui](https://user-images.githubusercontent.com/5365325/67035115-a00b1e80-f0de-11e9-9737-f3161891ac84.png)

[I've written a test that attempts to assert that the action button was clicked.](https://github.com/dv297/terra-react-testing-library/blob/master/src/SampleComponent.test.jsx#L14) But I noticed that the DOM never shows any of the content of the terra-alert. Here is what is outputted by the Jest console.

![dom](https://user-images.githubusercontent.com/5365325/67035339-1d369380-f0df-11e9-98e7-3b0dd3cd89c6.png)

You can see here that the text "Here is some text" is visible, so the SampleComponent is rendering, but the terra-alert just shows up as an empty div. I'm not totally sure why, and various other Terra components in my experience have rendered just fine in React testing library. I have a suspicion around the terra-responsive-element used by terra-alert's implementation, but I haven't dug too deep into it.

Even doing a "toMatchSnapshot" with React testing library does not render  the alert button. In contrast, doing the same using Enzyme's mount does render the button in the snapshot. 

So I'm not sure if it's a Terra issue or a React Testing Library issue, or maybe I'm misunderstanding React Testing Library since I'm new to it, but thought I'd pass it along for investigation purposes.

**Comment by cody-dot-js** [comment]
Yeah, this seems like an issue with the ResponsiveElement for sure. I've seen problems with using the `size` props (e.g. `tiny`, `small`, etc) before, hence this issue: https://github.com/cerner/terra-core/issues/2690

Curious what would happen if we switched the terra-alert implemenation to use the controlled version of the ResponsiveElement

**Comment by TheSavageDev** [comment]
Back when the discussion on this was occurring in Teams I went to try to refactor a small project we had in the Terra/Orion space. I was able to completely remove enzyme and solely test with testing-library. These tests cover: components with context, hooks, routers, etc. Got a pretty decent jump in code coverage too with just refactoring current tests.
